src/astral/page-script.js

```diff
@@ -0,0 +1,200 @@
+        const token = await getAccessToken();
+        const character = await getCharacter(characterName);
+
+        const recipients = whisper ? Object.fromEntries(getDungeonMasters().map(key => [key, true])) : undefined;
+        if (message.length > 2048) {
+            message = message.slice(0, 2045) + '...';
+        }
+        
+        return fetch(location.origin + `/api/game/${room}/chat`, {
+            method: "PUT",
+            body: JSON.stringify({
+                text: message,
+                color, 
+                icon, 
+                user, 
+                character, 
+                title: character || !characterName ? title : `${title} (${characterName})`,
+                hidden: whisper,
+                recipients
+            }),
+            headers: {
+                'x-authorization': `Bearer ${token}`, 'content-type': 'application/json'
+            }
+        });
+    } catch (e) {
+        console.warn(e);
+        message = `**${title}**\n\n` + message;
+
+        if (message.length > 2048) {
+            message = message.slice(0, 2045) + '...';
+        }
+        
+        await speakAs(characterName);
+        sendChatText(message);
+    }
+}
+
+async function updateHpBar({characterName, hp, maxHp, tempHp}) {
+    const room = getRoom();
+    const token = await getAccessToken();
+    const character = await getCharacter(characterName);
+    if (!character) {
+        console.error(`No character found with name ${characterName}`)
+        return
+    }
+    return fetch(location.origin + `/api/game/${room}/character/${character}`, {
+        method: "PATCH",
+        body: JSON.stringify({
+            character: {
+                updateAt: Date.now(),
+                attributeBarMax: maxHp + tempHp,
+                attributeBarCurrent: hp + tempHp,
+
+            }
+        }),
+        headers: {
+            'x-authorization': `Bearer ${token}`, 'content-type': 'application/json'
+        }
+    });
+}
+
+function disconnectAllEvents() {
+    for (let event of registered_events)
+        document.removeEventListener(...event);
+}
+
+var registered_events = [];
+registered_events.push(addCustomEventListener("AstralUpdateHPBar", updateHpBar));
+registered_events.push(addCustomEventListener("AstralChatMessage", postChatMessage));
+registered_events.push(addCustomEventListener("AstralRenderedRoll", handleRenderedRoll));
+registered_events.push(addCustomEventListener("disconnect", disconnectAllEvents));
+
+let chatIframe;
+
+function trySetDOMListeners() {
+    if (window.$ && $("div[data-id=chat] iframe").length > 0) {
+        chatIframe = $("div[data-id=chat] iframe").contents();
+        
+        // Added this since previews links will not work and they open a new tab
+        chatIframe.on('click', `a[href*='#beyond20-rendered-roll-button-']`, (ev) => {
+            ev.preventDefault();
+        })
+    } else {
+        setTimeout(trySetDOMListeners, 1000);
+    }
+}
```

Comment:
Move the global var and the function to the top of the file since `chatIframe` is used by the functions at the very top but doesn't get defined until the end, which was confusing during review. (the call to `trySetDOMListeners` should remain at the end of the file though)